include/ginkgo/core/base/mpi.hpp

```diff
@@ -503,9 +511,11 @@ class communicator {
      *                   MPI_Datatype.
      */
     template <typename SendType>
-    void send(const SendType* send_buffer, const int send_count,
-              const int destination_rank, const int send_tag) const
+    void send(std::shared_ptr<const Executor> exec, const SendType* send_buffer,
```

Comment:
The same story as before, now the function documentations are the ones which need to add the `exec` as `@param`. 

Also, it might make sense to give an `@note` in the `communicator` class to explain that communicator functions tied to a memory location require a `gko::Executor` to represent that, on top of the usual MPI parameters.